{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **DisplaCy visualiser tool for the metadata annotations**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import json\n",
    "from spacy import displacy\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Currently The annotations structure is as follows:\n",
    "\n",
    "\n",
    "```python\n",
    "start_index = int # The index of the first character of the entity. \n",
    "end_index = int # The index of the last character of the entity. (not inclusive)\n",
    "label = str # Label attached to the entity. \n",
    "\n",
    "{\n",
    "    \"classes\": [\"TEMP\", \"SOFT\", \"STIME\", \"MOL\", \"FFM\"],\n",
    "    \"annotations\": [\n",
    "        [\n",
    "            \"text_to_fill_in\",\n",
    "            {\n",
    "                \"entities\": [\n",
    "                    [start_index, end_index, label],\n",
    "                ]\n",
    "            }\n",
    "        ]\n",
    "    ]\n",
    "}\n",
    "```\n",
    "\n",
    "\n",
    "However, the spaCy  format for `ent` is as follows:\n",
    "\n",
    "```python\n",
    "start_index = int # The index of the first character of the entity. \n",
    "end_index = int # The index of the last character of the entity. (not inclusive)\n",
    "label = str # Label attached to the entity. \n",
    "\n",
    "{\n",
    "  \"text\": \"But Google is starting from behind.\",\n",
    "  \"ents\": [{ \"start\": start_index, \"end\": end_index, \"label\": label }]\n",
    "}\n",
    "\n",
    "# Example:\n",
    "{\n",
    "  \"text\": \"text_to_fill_in\",\n",
    "  \"ents\": [{ \"start\": 4, \"end\": 10, \"label\": \"ORG\" }]\n",
    "}\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "We need to therefore convert the format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_annotations(file_path):\n",
    "    \"\"\"\n",
    "    Convert annotations from the custom format to spaCy's displaCy format.\n",
    "    \n",
    "    Parameters:\n",
    "        file_path (str): Path to the JSON file containing \"classes\" and \"annotations\" keys.\n",
    "        \n",
    "    Returns:\n",
    "        List[dict]: A list of dictionaries in the spaCy format.\n",
    "    \"\"\"\n",
    "    with open(file_path, 'r') as file:\n",
    "        data = json.load(file)\n",
    "    \n",
    "    converted = []\n",
    "    for item in data[\"annotations\"]:\n",
    "        # Each item is structured as [text, {\"entities\": [[start, end, label], ...]}]\n",
    "        text, annotation_info = item\n",
    "        entities = annotation_info.get(\"entities\", [])\n",
    "        # Convert each entity to the spaCy dictionary format\n",
    "        ents = [{\"start\": start, \"end\": end, \"label\": label} for start, end, label in entities]\n",
    "        converted.append({\n",
    "            \"text\": text,\n",
    "            \"ents\": ents\n",
    "        })\n",
    "    return converted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'text': 'Molecular Dynamics Simulations of Sheet-like Replicators\\nThis dataset contains 3D structures, MD trajectories and powder X-ray diffractograms (simulated and experimental) of tetrameric carboxynaphthalene disulfide. The MD simulations were performed in water with periodic boundary conditions. Either AMBER or GROMACS packages were used. For detailed description of the simulations see the linked paper (the link will appear upon acceptance). PDB files are snapshots from the MD simulations; they can be viewed using any 3D structure viewer (e.g. Avogadro). The trajectories (*.xtc - GROMACS, *.mdcrd - AMBER/NetCDF) can be visualized using e.g. VMD. Partial charges for the naphthalene building blocks are provided within *.mol2 files (N-0 for the neutral and N-1 for the deprotonated subunit). The parameter/topology files have .top (GROMACS) and .prmtop (AMBER) extensions. Files labeled \"sheet1\" contain results of simulations of 16 repeats of tetramer IV showed in Figure 4 and S13 of the paper. The suffix \"-exp box\" refers to structures which were constrained to the box determined experimentally (\"min\" - after energy minimization, \"equil\" - after equilibration). Files labeled \"sheet2\" contain results of simulations of 12 repeats of tetramer IV showed in Figure S14 of the paper. Files labeled \"sheet3\" contain results of simulations of 16 repeats of tetramer IV showed in Figure S15 of the paper. The assembly was unstable, the initial structure started dissociating from the very beginning of the simulation. The trajectory is in NetCDF format. Files labeled \"tetramerIV-2\" contain results of simulations of isolated, partially deprotonated tetramer IV in water (solvent was removed from the trajectory). Link to the article will be added upon its acceptance.', 'ents': [{'start': 185, 'end': 213, 'label': 'MOL'}, {'start': 252, 'end': 257, 'label': 'MOL'}, {'start': 300, 'end': 305, 'label': 'SOFT'}, {'start': 309, 'end': 316, 'label': 'SOFT'}, {'start': 583, 'end': 590, 'label': 'SOFT'}, {'start': 602, 'end': 607, 'label': 'SOFT'}, {'start': 645, 'end': 648, 'label': 'SOFT'}, {'start': 835, 'end': 842, 'label': 'SOFT'}, {'start': 857, 'end': 862, 'label': 'SOFT'}, {'start': 1667, 'end': 1672, 'label': 'MOL'}]}]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Molecular Dynamics Simulations of Sheet-like Replicators<br>This dataset contains 3D structures, MD trajectories and powder X-ray diffractograms (simulated and experimental) of tetrameric \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    carboxynaphthalene disulfide\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ". The MD simulations were performed in \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " with periodic boundary conditions. Either \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    AMBER\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       " or \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       " packages were used. For detailed description of the simulations see the linked paper (the link will appear upon acceptance). PDB files are snapshots from the MD simulations; they can be viewed using any 3D structure viewer (e.g. Avogadro). The trajectories (*.xtc - \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       ", *.mdcrd - \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    AMBER\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       "/NetCDF) can be visualized using e.g. \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    VMD\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       ". Partial charges for the naphthalene building blocks are provided within *.mol2 files (N-0 for the neutral and N-1 for the deprotonated subunit). The parameter/topology files have .top (\n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       ") and .prmtop (\n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    AMBER\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       ") extensions. Files labeled &quot;sheet1&quot; contain results of simulations of 16 repeats of tetramer IV showed in Figure 4 and S13 of the paper. The suffix &quot;-exp box&quot; refers to structures which were constrained to the box determined experimentally (&quot;min&quot; - after energy minimization, &quot;equil&quot; - after equilibration). Files labeled &quot;sheet2&quot; contain results of simulations of 12 repeats of tetramer IV showed in Figure S14 of the paper. Files labeled &quot;sheet3&quot; contain results of simulations of 16 repeats of tetramer IV showed in Figure S15 of the paper. The assembly was unstable, the initial structure started dissociating from the very beginning of the simulation. The trajectory is in NetCDF format. Files labeled &quot;tetramerIV-2&quot; contain results of simulations of isolated, partially deprotonated tetramer IV in \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " (solvent was removed from the trajectory). Link to the article will be added upon its acceptance.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "data = \"../annotations/figshare_1381865.json\"\n",
    "\n",
    "# Convert the annotations.\n",
    "converted_data = convert_annotations(data)\n",
    "print(converted_data)\n",
    "\n",
    "colors = {\n",
    "    \"TEMP\": \"#ffb3ba\",\n",
    "    \"SOFT\": \"#ffffba\",\n",
    "    \"STIME\": \"#baffc9\",\n",
    "    \"MOL\": \"#bae1ff\",\n",
    "    \"FFM\": \"#cdb4db\",\n",
    "}\n",
    "options = {\"colors\": colors}\n",
    "\n",
    "# You now have a list of annotation dictionaries. For visualization with displaCy,\n",
    "# you can serve a single annotated text or multiple.\n",
    "displacy.render(converted_data, style=\"ent\", manual=True, options=options)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **We are going to be doing some prompt engineering**\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These are just functions to convert the AI outputs to a strcuture compatible with spaCy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_annotated_text_to_spacy_format(annotated_text):\n",
    "\n",
    "    with open(annotated_text, 'r') as file:\n",
    "        annotated_text = file.read()\n",
    "\n",
    "    pattern = re.compile(r'<(MOL|STIME|TEMP|SOFTNAME|SOFTVERS|FFM)>(.+?)<\\1>')\n",
    "    ents = []\n",
    "    clean_text = \"\"\n",
    "    last_idx = 0  # Position in the annotated text\n",
    "\n",
    "    for match in pattern.finditer(annotated_text):\n",
    "        tag = match.group(1)\n",
    "        entity_text = match.group(2)\n",
    "\n",
    "        # Text between last entity and current\n",
    "        pre_entity_text = annotated_text[last_idx:match.start()]\n",
    "        clean_text += pre_entity_text\n",
    "        entity_start = len(clean_text)\n",
    "        clean_text += entity_text\n",
    "        entity_end = len(clean_text)\n",
    "\n",
    "        ents.append({\n",
    "            \"start\": entity_start,\n",
    "            \"end\": entity_end,\n",
    "            \"label\": tag\n",
    "        })\n",
    "\n",
    "        last_idx = match.end()\n",
    "\n",
    "    # Add any remaining text after last tag\n",
    "    clean_text += annotated_text[last_idx:]\n",
    "\n",
    "    return [{\n",
    "        \"text\": clean_text,\n",
    "        \"ents\": ents\n",
    "    }]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def visualize_annotated_texts(data:list)->None:\n",
    "    \"\"\"\n",
    "    Visualize the annotated text using displaCy.\n",
    "    \n",
    "    Parameters:\n",
    "        data (list): List of paths to the annotated text files.\n",
    "    \"\"\"\n",
    "   # Convert the annotations.\n",
    "    converted_real_data = convert_annotations(data[0]) # Real annotation\n",
    "    converted_ai_data = convert_annotated_text_to_spacy_format(data[1]) # AI annotated\n",
    "\n",
    "    colors = {\n",
    "        \"TEMP\": \"#ffb3ba\",\n",
    "        \"SOFTNAME\": \"#ffffba\",\n",
    "        \"SOFTVERS\": \"#orange\",\n",
    "        \"STIME\": \"#baffc9\",\n",
    "        \"MOL\": \"#bae1ff\",\n",
    "        \"FFM\": \"#cdb4db\",\n",
    "    }\n",
    "    options = {\"colors\": colors}\n",
    "\n",
    "    print(\"\\nREAL ANNOTATION:\")\n",
    "    displacy.render(converted_real_data, style=\"ent\", manual=True, options=options)\n",
    "    print(\"\\nAI ANNOTATION:\")\n",
    "    displacy.render(converted_ai_data, style=\"ent\", manual=True, options=options)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Let's use multiple different texts with ChatGPT4o - Detailed few-shot propmting**\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```txt\n",
    "The aim is to extract entities from an abstract or description of a molecular dynamics dataset. The following entities will be ones to look out for:\n",
    "\n",
    "- software name (SOFTNAME): Name of any software identified\n",
    "- software version (SOFTVERS): Version associated with the software\n",
    "- molecule (MOL): Gene, protein, molecule-derived entity present in the simulation\n",
    "- simulation time (STIME): Simulation time\n",
    "- temperature (TEMP): Temperature used in the simulation\n",
    "- force field (FFM): Force field used for the simulation\n",
    "\n",
    "Here are some examples of the entities:\n",
    "\n",
    "- software name (SOFTNAME): gromacs, mdtraj, charmm-gui, namd, mdanalysis, glide\n",
    "- software version (SOFTVERS): 3.0.0, 24.1,  14.0\n",
    "- molecule (MOL): POPC, cholesterol, ubiquitin, TSPO, kinase, phenytoin, water, Hp\n",
    "- simulation time (STIME): 2 ns, 5 µs, 100 ns\n",
    "- temperature (TEMP): 300 K, 310.15 K, 280 K, 298 K\n",
    "- force field (FFM): charmm36m, martini, amber\n",
    "\n",
    "Here are a couple of examples of abstracts, followed by their proper, correct annotation.\n",
    "\n",
    "Example 1\n",
    "Abstract: \n",
    "\"POPC Ulmschneider OPLS NaCl 1Mol\\nMD simulation trajectory and related files for fully hydrated POPC bilayer with 1M NaCl. The Ulmschneider force field for POPC was used with Gromacs 5.0.3 [1,2]. Ions were described by OPLS-AA force field (Gromacs 5.0.3). Conditions: T 298.15, 128 POPC molecules, 5120 tip3p waters, 77 Na, 77 Cl. 200ns trajectory (preceded by 5ns NPT equillibration). This data is ran for the nmrlipids.blospot.fi project. More details from nmrlipids.blospot.fi and https://removed [1] J.P. Ulmschneider & M.B. Ulmschneider, United Atom Lipid Parameters for Combination with the Optimized Potentials for Liquid Simulations All-Atom Force Field, JCTC 2009, 5 (7), 1803 1813 [2] https://removed\"\n",
    "\n",
    "Annotation: \n",
    "{\"classes\": [\"TEMP\", \"SOFTNAME\", “SOFTVERS”, \"STIME\", \"MOL\", \"FFM\"], \"annotations\": [[\"POPC Ulmschneider OPLS NaCl 1Mol\\nMD simulation trajectory and related files for fully hydrated POPC bilayer with 1M NaCl. The Ulmschneider force field for POPC was used with Gromacs 5.0.3 [1,2]. Ions were described by OPLS-AA force field (Gromacs 5.0.3). Conditions: T 298.15, 128 POPC molecules, 5120 tip3p waters, 77 Na, 77 Cl. 200ns trajectory (preceded by 5ns NPT equillibration). This data is ran for the nmrlipids.blospot.fi project. More details from nmrlipids.blospot.fi and https://removed [1] J.P. Ulmschneider & M.B. Ulmschneider, United Atom Lipid Parameters for Combination with the Optimized Potentials for Liquid Simulations All-Atom Force Field, JCTC 2009, 5 (7), 1803 1813 [2] https://removed\", {\"entities\": [[0, 4, \"MOL\"], [5, 17, \"FFM\"], [18, 22, \"FFM\"], [23, 27, \"MOL\"], [95, 99, \"MOL\"], [116, 120, \"MOL\"], [126, 138, \"FFM\"], [155, 159, \"MOL\"], [174, 181, \"SOFTNAME\"], [182, 187, \"SOFTVERS\"], [195, 199, \"MOL\"], [218, 225, \"FFM\"], [239, 246, \"SOFTNAME\"], [247, 252, \"SOFTVERS\"], [269, 275, \"TEMP\"], [281, 285, \"MOL\"], [302, 307, \"FFM\"], [308, 314, \"MOL\"], [319, 321, \"MOL\"], [326, 328, \"MOL\"], [330, 335, \"STIME\"]]}]]}\n",
    "\n",
    "Example 2\n",
    "Abstract: \n",
    "\"6 molecular dynamics simulations of coronavirus 2019-nCoV protease model in complex with different conformations of lopinavir.\\nMolecular dynamics simulations of comparative model of novel coronavirus 2019-nCoV protease Mpro in complex with 6 different conformations based on the Catalophore point-cloud alignment and (re)- docking of lopinavir into the 2019ncov virus protease model. The docking experiment produced 8 clusters of possible conformations, we chose 6 out of 8 conformers and ran an all-atom 300 ps MD at 310 K ( 36.85 C ).-The two images in the main folder refers to the docked structures before MD simulation.-The file all centroids.pse contains the frames representing the centroid of the subsequent MD simulation for each docking cluster.-Each archive contains the centroid in PDB format, the starting frame of the simulation in GRO format and the compressed trajectory in XTC format. In the directory \\\"other files\\\" there are other data generated during the simulation, i.e. heatmap representing the contact frequency between the ligand atoms and the ones belonging to the homology model.\"\n",
    "\n",
    "Annotation: \n",
    "{\"classes\": [\"TEMP\", \"SOFTNAME\", “SOFTVERS”, \"STIME\", \"MOL\", \"FFM\"], \"annotations\": [[\"6 molecular dynamics simulations of coronavirus 2019-nCoV protease model in complex with different conformations of lopinavir.\\nMolecular dynamics simulations of comparative model of novel coronavirus 2019-nCoV protease Mpro in complex with 6 different conformations based on the Catalophore point-cloud alignment and (re)- docking of lopinavir into the 2019ncov virus protease model. The docking experiment produced 8 clusters of possible conformations, we chose 6 out of 8 conformers and ran an all-atom 300 ps MD at 310 K ( 36.85 C ).-The two images in the main folder refers to the docked structures before MD simulation.-The file all centroids.pse contains the frames representing the centroid of the subsequent MD simulation for each docking cluster.-Each archive contains the centroid in PDB format, the starting frame of the simulation in GRO format and the compressed trajectory in XTC format. In the directory \\\"other files\\\" there are other data generated during the simulation, i.e. heatmap representing the contact frequency between the ligand atoms and the ones belonging to the homology model.\", {\"entities\": [[36, 66, \"MOL\"], [116, 125, \"MOL\"], [188, 218, \"MOL\"], [219, 223, \"MOL\"], [334, 343, \"MOL\"], [353, 376, \"MOL\"], [518, 523, \"TEMP\"], [526, 533, \"TEMP\"]]}]]}\n",
    "\n",
    "I will want you to annotate with the following structure making sure to keep using \\n for line breaks. Don’t manually go to the line. Make sure to only answer with the output:\n",
    "“text <ENTITY LABEL>entity_text<ENTITY LABEL> text”\n",
    "\n",
    "Example:\n",
    "“<MOL>POPC<MOL> >FFM>Ulmschneider<FFM> <FFM>OPLS<FFM> <MOL>NaCl<MOL> 1Mol”\n",
    "\n",
    "Here is an abstract, and I need you to properly annotate it with the entities I explained before:\n",
    "[INSERT ABSTRACT]\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "REAL ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Molecular Dynamics Simulations of Sheet-like Replicators<br>This dataset contains 3D structures, MD trajectories and powder X-ray diffractograms (simulated and experimental) of tetrameric \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    carboxynaphthalene disulfide\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ". The MD simulations were performed in \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " with periodic boundary conditions. Either \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    AMBER\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       " or \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       " packages were used. For detailed description of the simulations see the linked paper (the link will appear upon acceptance). PDB files are snapshots from the MD simulations; they can be viewed using any 3D structure viewer (e.g. Avogadro). The trajectories (*.xtc - \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       ", *.mdcrd - \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    AMBER\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       "/NetCDF) can be visualized using e.g. \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    VMD\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       ". Partial charges for the naphthalene building blocks are provided within *.mol2 files (N-0 for the neutral and N-1 for the deprotonated subunit). The parameter/topology files have .top (\n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       ") and .prmtop (\n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    AMBER\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       ") extensions. Files labeled &quot;sheet1&quot; contain results of simulations of 16 repeats of tetramer IV showed in Figure 4 and S13 of the paper. The suffix &quot;-exp box&quot; refers to structures which were constrained to the box determined experimentally (&quot;min&quot; - after energy minimization, &quot;equil&quot; - after equilibration). Files labeled &quot;sheet2&quot; contain results of simulations of 12 repeats of tetramer IV showed in Figure S14 of the paper. Files labeled &quot;sheet3&quot; contain results of simulations of 16 repeats of tetramer IV showed in Figure S15 of the paper. The assembly was unstable, the initial structure started dissociating from the very beginning of the simulation. The trajectory is in NetCDF format. Files labeled &quot;tetramerIV-2&quot; contain results of simulations of isolated, partially deprotonated tetramer IV in \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " (solvent was removed from the trajectory). Link to the article will be added upon its acceptance.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Molecular Dynamics Simulations of Sheet-like Replicators\\nThis dataset contains 3D structures, MD trajectories and powder X-ray diffractograms (simulated and experimental) of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    tetrameric carboxynaphthalene disulfide\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ". The MD simulations were performed in \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " with periodic boundary conditions. Either \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    AMBER\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " or \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " packages were used. For detailed description of the simulations see the linked paper (the link will appear upon acceptance). PDB files are snapshots from the MD simulations; they can be viewed using any 3D structure viewer (e.g. \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Avogadro\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       "). The trajectories (*.xtc - \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       ", *.mdcrd - \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    AMBER\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       "/NetCDF) can be visualized using e.g. \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    VMD\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       ". Partial charges for the \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    naphthalene\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " building blocks are provided within *.mol2 files (N-0 for the neutral and N-1 for the deprotonated subunit). The parameter/topology files have .top (\n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       ") and .prmtop (\n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    AMBER\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       ") extensions. Files labeled &quot;sheet1&quot; contain results of simulations of 16 repeats of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    tetramer IV\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " showed in Figure 4 and S13 of the paper. The suffix &quot;-exp box&quot; refers to structures which were constrained to the box determined experimentally (&quot;min&quot; - after energy minimization, &quot;equil&quot; - after equilibration). Files labeled &quot;sheet2&quot; contain results of simulations of 12 repeats of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    tetramer IV\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " showed in Figure S14 of the paper. Files labeled &quot;sheet3&quot; contain results of simulations of 16 repeats of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    tetramer IV\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " showed in Figure S15 of the paper. The assembly was unstable, the initial structure started dissociating from the very beginning of the simulation. The trajectory is in NetCDF format. Files labeled &quot;\n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    tetramerIV-2\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "&quot; contain results of simulations of isolated, partially deprotonated \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    tetramer IV\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " in \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " (solvent was removed from the trajectory). Link to the article will be added upon its acceptance.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "real_data = \"../annotations/figshare_1381865.json\"\n",
    "ai_data = \"../ai_annotations/chatgpt_annotated.txt\"\n",
    "\n",
    "data = [real_data, ai_data]\n",
    "\n",
    "visualize_annotated_texts(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "REAL ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Classification of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-dependent \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    K-Ras4B\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " active and inactive conformational states<br>Dataset for the paper: Classification of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-dependent \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    K-Ras4B\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " active and inactive conformational states. Cite as: J. Chem. Phys. 158, 000000 (2023); DOI: 10.1063/5.0139181 Submitted: 18 December 2022; Accepted: 13 February 2023; Published Online: 13 February 2023. All molecular dynamics and molecular docking data presented, analyzed, and discussed in this paper are available at reasonable requests submitted to the corresponding author. The following data are also available online: (i) file KRas4B pdbs.zip, a compressed archive including coordinate (.pdb) and structure (.psf) files for \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRas-4B\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " WT and \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    D33E\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    proteins\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ", (ii) KRas4B WT traj.zip, a compressed archive including trajectory files (.trr) for the WT \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRas-4B\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " runs, including 120 &quot;*.trr&quot;-formatted trajectories each corresponding to \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    40 ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       " of MD simulation time, and (iii) a sample Python script to generate a free energy plot as shown in Fig. 2.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Classification of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-dependent \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    K-Ras4B\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " active and inactive conformational states\\nDataset for the paper: Classification of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-dependent \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    K-Ras4B\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " active and inactive conformational states. Cite as: J. Chem. Phys. 158, 000000 (2023); DOI: 10.1063/5.0139181 Submitted: 18 December 2022; Accepted: 13 February 2023; Published Online: 13 February 2023. All molecular dynamics and molecular docking data presented, analyzed, and discussed in this paper are available at reasonable requests submitted to the corresponding author. The following data are also available online: (i) file \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRas4B\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " pdbs.zip, a compressed archive including coordinate (.pdb) and structure (.psf) files for \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRas-4B\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    WT\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    D33E\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " proteins, (ii) \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRas4B\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    WT\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " traj.zip, a compressed archive including trajectory files (.trr) for the \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    WT\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRas-4B\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " runs, including 120 &quot;*.trr&quot;-formatted trajectories each corresponding to \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    40 ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       " of MD simulation time, and (iii) a sample Python script to generate a free energy plot as shown in Fig. 2.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "real_data = \"../annotations/zenodo_7672195.json\"\n",
    "ai_data = \"../ai_annotations/chatgpt_zenodo_7672195.txt\"\n",
    "\n",
    "data = [real_data, ai_data]\n",
    "\n",
    "visualize_annotated_texts(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "REAL ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">An in silico approach to determine inter-subunit affinities in human \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    septin\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " complexes<br>The \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    septins\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " are a conserved family of filament-forming guanine nucleotide binding proteins, often named the fourth component of the cytoskeleton. Correctly assembled \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    septin\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " structures are required for essential intracellular processes such as cytokinesis, vesicular transport, polarity establishment, and cellular adhesion. \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Septins\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " structurally belong to the \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    P-Loop NTPases\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " but they do not mediate signals to effectors through \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " binding and hydrolysis. \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " binding and hydrolysis are believed to contribute to \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    septin\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " complex integrity, but biochemical approaches addressing this topic are hampered by the stability of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    septin\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " complexes after recombinant expression and the lack of nucleotide-depleted complexes. To complement this limitation, we used a molecular dynamics based approach to determine inter-subunit free binding energies in available human \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    septin\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " dimer structures and in their apo forms, which we generated in silico. The \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    nucleotide\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " in the \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTPase\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " active subunits \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    SEPT2\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    SEPT7\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ", but not in \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    SEPT6\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ", was identified as a stabilizing element in the G interface as it is coordinated at its ribose ring to conserved amino acids. Removal of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GDP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " from \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    SEPT2\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    SEPT7\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " results in flipping of a conserved Arg residue and disruption of an extensive hydrogen bond network in the \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    septin\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " unique element, concomitant with a decreased inter-subunit affinity.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">An in silico approach to determine inter-subunit affinities in human \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    septin complexes\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "\\nThe septins are a conserved family of filament-forming \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    guanine nucleotide\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " binding proteins, often named the fourth component of the cytoskeleton. Correctly assembled septin structures are required for essential intracellular processes such as cytokinesis, vesicular transport, polarity establishment, and cellular adhesion. Septins structurally belong to the \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    P-Loop NTPases\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " but they do not mediate signals to effectors through \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " binding and hydrolysis. \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " binding and hydrolysis are believed to contribute to \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    septin complex\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " integrity, but biochemical approaches addressing this topic are hampered by the stability of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    septin complexes\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " after recombinant expression and the lack of nucleotide-depleted complexes. To complement this limitation, we used a molecular dynamics based approach to determine inter-subunit free binding energies in available human \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    septin dimer\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " structures and in their apo forms, which we generated in silico. The nucleotide in the GTPase active subunits \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    SEPT2\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    SEPT7\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ", but not in \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    SEPT6\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ", was identified as a stabilizing element in the G interface as it is coordinated at its ribose ring to conserved amino acids. Removal of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GDP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " from \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    SEPT2\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    SEPT7\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " results in flipping of a conserved Arg residue and disruption of an extensive hydrogen bond network in the septin unique element, concomitant with a decreased inter-subunit affinity.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "real_data = \"../annotations/zenodo_7440399.json\"\n",
    "ai_data = \"../ai_annotations/chatgpt_zenodo_7440399.txt\"\n",
    "\n",
    "data = [real_data, ai_data]\n",
    "\n",
    "visualize_annotated_texts(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "REAL ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">\n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    POPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " Ulmschneider \n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    OPLS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       " Verlet Group<br>MD simulation trajectory and related files for fully hydrated \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    POPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " bilayer run with Verlet and Group schemes. The Ulmschneider force field for \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    POPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " was used with \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Gromacs 5.0.3\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       " [1,2]. Conditions: T \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    298.15\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       ", 128 \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    POPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules, 5120 \n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    tip3p\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    waters\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ". \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    200ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       " trajectory (preceded by 5ns NPT equillibration). Starting structure was obtained from \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    CHARMM-GUI\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       " [3]. This data is ran for the nmrlipids.blospot.fi project. More details from nmrlipids.blospot.fi and https://removed [1] J.P. Ulmschneider &amp; M.B. Ulmschneider, United Atom Lipid Parameters for Combination with the Optimized Potentials for Liquid Simulations All-Atom Force Field, JCTC 2009, 5 (7), 1803 1813 [2] https://removed [3] https://removed</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">\n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    POPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Ulmschneider\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    OPLS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       " Verlet Group\\nMD simulation trajectory and related files for fully hydrated \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    POPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " bilayer run with Verlet and Group schemes. The \n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Ulmschneider\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       " force field for \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    POPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " was used with \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Gromacs\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #orange; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    5.0.3\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTVERS</span>\n",
       "</mark>\n",
       " [1,2]. Conditions: T \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    298.15\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       ", 128 \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    POPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules, 5120 \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    tip3p\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    waters\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ". \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    200ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       " trajectory (preceded by \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    5ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       " NPT equillibration). Starting structure was obtained from \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    CHARMM-GUI\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " [3]. This data is ran for the nmrlipids.blospot.fi project. More details from nmrlipids.blospot.fi and https://removed [1] J.P. Ulmschneider &amp; M.B. Ulmschneider, United Atom Lipid Parameters for Combination with the Optimized Potentials for Liquid Simulations All-Atom Force Field, JCTC 2009, 5 (7), 1803 1813 [2] https://removed [3] https://removed</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "real_data = \"../annotations/zenodo_30904.json\"\n",
    "ai_data = \"../ai_annotations/chatgpt_zenodo_30904.txt\"\n",
    "\n",
    "data = [real_data, ai_data]\n",
    "\n",
    "visualize_annotated_texts(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "REAL ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">212 \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    DPPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " Molecules bilayer in pure \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ", simulated at temperatures ranging from \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       " to \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    358K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       "<br>Publication: A machine learning study of the two states model for lipid bilayer phase transitions Published on: 12 August 2020 Journal: Phys. Chem. Chem. Phys., 2020, DOI: 10.1039/D0CP02058C Description: Simulation files used to train our machine learning algorithm to identify the thermodynamic phase of individual lipid molecules in a bilayer, as well as the simulation files analysed by the machine learning models. Code source for the ML algorithm can be found on Github. The training files are named gel.gro and fluid.gro. They respectively correspond to the final frame of the systems simulated at \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    358K.\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       " All other files are the files analysed by the machine learning models. System composition: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    DPPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules: 212 with 130 atoms each \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules: 29,826 with 3 atoms each Simulation box dimensions (approx.): 8 x 8 x 20 nm Simulation details: Software: \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Gromacs (v. 2016.4)\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       " Forcefield: \n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Charmm36 (v. June 2015)\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       " - \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ": \n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    TIP3P\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       " Thermostat: Nose-hoover (0.4ps, 2 groups) Barostat: Parrinello-Rahman semi-isotropic (2.0ps, 1.0 bar on each axis, 4.5e-5 bar-1) Duration: \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    50 ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       " </div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">212 \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    DPPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " Molecules bilayer in pure \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ", simulated at temperatures ranging from \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       " to \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    358K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       "\\nPublication: A machine learning study of the two states model for lipid bilayer phase transitions Published on: 12 August 2020 Journal: Phys. Chem. Chem. Phys., 2020, DOI: 10.1039/D0CP02058C Description: Simulation files used to train our machine learning algorithm to identify the thermodynamic phase of individual lipid molecules in a bilayer, as well as the simulation files analysed by the machine learning models. Code source for the ML algorithm can be found on Github. The training files are named gel.gro and fluid.gro. They respectively correspond to the final frame of the systems simulated at \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    358K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       ". All other files are the files analysed by the machine learning models. System composition: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    DPPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules: 212 with 130 atoms each \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules: 29,826 with 3 atoms each Simulation box dimensions (approx.): 8 x 8 x 20 nm Simulation details: Software: \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Gromacs\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " (\n",
       "<mark class=\"entity\" style=\"background: #orange; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    v. 2016.4\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTVERS</span>\n",
       "</mark>\n",
       ") Forcefield: \n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Charmm36\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       " (\n",
       "<mark class=\"entity\" style=\"background: #orange; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    June 2015\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTVERS</span>\n",
       "</mark>\n",
       ") - \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ": \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    TIP3P\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " Thermostat: Nose-hoover (0.4ps, 2 groups) Barostat: Parrinello-Rahman semi-isotropic (2.0ps, 1.0 bar on each axis, 4.5e-5 bar-1) Duration: \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    50 ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "real_data = \"../annotations/zenodo_3950029.json\"\n",
    "ai_data = \"../ai_annotations/chatgpt_zenodo_3950029.txt\"\n",
    "\n",
    "data = [real_data, ai_data]\n",
    "\n",
    "visualize_annotated_texts(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Let's try with a different prompt: few shot prompting**\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```txt\n",
    "Entities:\n",
    "<SOFTNAME>: Software used in the simulation or analysis (e.g., Gromacs, AMBER, VMD, CHARMM-GUI)\n",
    "<SOFTVERS>: Version number associated with the software (e.g., v. 2016.4, 5.0.3)\n",
    "<MOL>: Molecules, proteins, lipids, water models, or molecular complexes involved (e.g., DPPC, water, GTP, KRas4B)\n",
    "<STIME>: Duration of the simulation (e.g., 50 ns, 200ns, 5 µs)\n",
    "<TEMP>: Temperature used in the simulation (e.g., 300 K, 288K, 358K)\n",
    "<FFM>: Force fields used in the simulation (e.g., Charmm36, AMBER, MARTINI, TIP3P)\n",
    "\n",
    "Annotate the text:\n",
    "POPC Ulmschneider OPLS Verlet Group\\nMD simulation trajectory and related files for fully hydrated POPC bilayer run with Verlet and Group schemes. The Ulmschneider force field for POPC was used with Gromacs 5.0.3 [1,2]. Conditions: T 298.15, 128 POPC molecules, 5120 tip3p waters. 200ns trajectory (preceded by 5ns NPT equillibration). Starting structure was obtained from CHARMM-GUI [3]. This data is ran for the nmrlipids.blospot.fi project. More details from nmrlipids.blospot.fi and https://removed [1] J.P. Ulmschneider & M.B. Ulmschneider, United Atom Lipid Parameters for Combination with the Optimized Potentials for Liquid Simulations All-Atom Force Field, JCTC 2009, 5 (7), 1803 1813 [2] https://removed [3] https://removed\n",
    "\n",
    "Annotated text:\n",
    "<MOL>POPC<MOL> <FFM>Ulmschneider<FFM> <FFM>OPLS<FFM> Verlet Group\\nMD simulation trajectory and related files for fully hydrated <MOL>POPC<MOL> bilayer run with Verlet and Group schemes. The <FFM>Ulmschneider<FFM> force field for <MOL>POPC<MOL> was used with <SOFTNAME>Gromacs<SOFTNAME> <SOFTVERS>5.0.3<SOFTVERS> [1,2]. Conditions: T <TEMP>298.15<TEMP>, 128 <MOL>POPC<MOL> molecules, 5120 <MOL>tip3p<MOL> <MOL>waters<MOL>. <STIME>200ns<STIME> trajectory (preceded by <STIME>5ns<STIME> NPT equillibration). Starting structure was obtained from <SOFTNAME>CHARMM-GUI<SOFTNAME> [3]. This data is ran for the nmrlipids.blospot.fi project. More details from nmrlipids.blospot.fi and https://removed [1] J.P. Ulmschneider & M.B. Ulmschneider, United Atom Lipid Parameters for Combination with the Optimized Potentials for Liquid Simulations All-Atom Force Field, JCTC 2009, 5 (7), 1803 1813 [2] https://removed [3] https://removed\n",
    "\n",
    "\n",
    "Annotate the text:\n",
    "Assessment of mutation probabilities of KRAS G12 missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling: Datasets.\\nDatasets related to the publication [1]. Including: KRAS G12X mutations derived from COSMIC v.79 [http://cancer.sanger.ac.uk/cosmic/] (KRAS G12X mut COSMICv79..xlsx) RMSFs (300-2000ns) of GDP-systems (300 2000rmsf GDP systems RAW AVG SE.xlsx) RMSFs (300-2000ns) of GTP-systems (300 2000RMSF GTP systems RAW AVG SE.xlsx) PyInteraph analysis data for salt-bridges and hydrophobic clusters (.dat files for each system in the PyInteraph data.zip-file) Backbone trajectories for each system (residues 4-164; frames for every 1ns). Last number (e.g. 1) refers to the replica of the simulated system. backbone 4-164.gro/.pdb/.tpr -files (resid 4-164) [1] Pantsar T et al. Assessment of mutation probabilities of KRAS G12 missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling. PLoS Comput Biol Submitted (2018)\n",
    "\n",
    "Annotated text:\n",
    "Assessment of mutation probabilities of <MOL>KRAS G12<MOL> missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling: Datasets.\\nDatasets 1  related to the publication [1]. Including: <MOL>KRAS G12X<MOL> mutations derived from COSMIC v.79 [http://cancer.sanger.ac.uk/cosmic/] (<MOL>KRAS G12X<MOL> mut COSMICv79..xlsx) RMSFs (<STIME>300-2000ns<STIME>) of <MOL>GDP<MOL>-systems (<STIME>300-2000ns<STIME> rmsf <MOL>GDP<MOL> systems RAW AVG SE.xlsx) RMSFs (<STIME>300-2000ns<STIME>) of <MOL>GTP<MOL>-systems (<STIME>300-2000ns<STIME> RMSF <MOL>GTP<MOL> systems RAW AVG SE.xlsx) <SOFTNAME>PyInteraph<SOFTNAME> analysis data for salt-bridges and hydrophobic clusters (.dat files for each system in the <SOFTNAME>PyInteraph<SOFTNAME> data.zip-file) Backbone trajectories for each system (residues 4-164; frames for every <STIME>1ns<STIME>). Instead of using <SOFTNAME>GROMACS<SOFTNAME>, we used <SOFTNAME>NAMD<SOFTNAME> also\n",
    "\n",
    "\n",
    "Annotate the text:\n",
    "212 DPPC Molecules bilayer in pure Water, simulated at temperatures ranging from 288K to 358K\\nPublication: A machine learning study of the two states model for lipid bilayer phase transitions Published on: 12 August 2020 Journal: Phys. Chem. Chem. Phys., 2020, DOI: 10.1039/D0CP02058C Description: Simulation files used to train our machine learning algorithm to identify the thermodynamic phase of individual lipid molecules in a bilayer, as well as the simulation files analysed by the machine learning models. Code source for the ML algorithm can be found on Github. The training files are named gel.gro and fluid.gro. They respectively correspond to the final frame of the systems simulated at 288K and 358K. All other files are the files analysed by the machine learning models. System composition: DPPC molecules: 212 with 130 atoms each Water molecules: 29,826 with 3 atoms each Simulation box dimensions (approx.): 8 x 8 x 20 nm Simulation details: Software: Gromacs (v. 2016.4) Forcefield: Charmm36 (v. June 2015) - Water: TIP3P Thermostat: Nose-hoover (0.4ps, 2 groups) Barostat: Parrinello-Rahman semi-isotropic (2.0ps, 1.0 bar on each axis, 4.5e-5 bar-1) Duration: 50 ns\n",
    "\n",
    "Annotated text:\n",
    "\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION - chatgpt4o:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">212 \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    DPPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " Molecules bilayer in pure \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ", simulated at temperatures ranging from \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       " to \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    358K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       "\\nPublication: A machine learning study of the two states model for lipid bilayer phase transitions Published on: 12 August 2020 Journal: Phys. Chem. Chem. Phys., 2020, DOI: 10.1039/D0CP02058C Description: Simulation files used to train our machine learning algorithm to identify the thermodynamic phase of individual lipid molecules in a bilayer, as well as the simulation files analysed by the machine learning models. Code source for the ML algorithm can be found on Github. The training files are named gel.gro and fluid.gro. They respectively correspond to the final frame of the systems simulated at \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    358K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       ". All other files are the files analysed by the machine learning models. System composition: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    DPPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules: 212 with 130 atoms each \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules: 29,826 with 3 atoms each Simulation box dimensions (approx.): 8 x 8 x 20 nm Simulation details: Software: \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Gromacs\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " (\n",
       "<mark class=\"entity\" style=\"background: #orange; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    v. 2016.4\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTVERS</span>\n",
       "</mark>\n",
       ") Forcefield: \n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Charmm36\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       " (\n",
       "<mark class=\"entity\" style=\"background: #orange; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    v. June 2015\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTVERS</span>\n",
       "</mark>\n",
       ") - \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ": \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    TIP3P\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " Thermostat: Nose-hoover (0.4ps, 2 groups) Barostat: Parrinello-Rahman semi-isotropic (2.0ps, 1.0 bar on each axis, 4.5e-5 bar-1) Duration: \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    50 ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION - gemini_2_0_flash:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">\n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    212 DPPC Molecules\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " bilayer in pure \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ", simulated at temperatures ranging from \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       " to \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    358K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       "\\nPublication: A machine learning study of the two states model for lipid bilayer phase transitions Published on: 12 August 2020 Journal: Phys. Chem. Chem. Phys., 2020, DOI: 10.1039/D0CP02058C Description: Simulation files used to train our machine learning algorithm to identify the thermodynamic phase of individual lipid molecules in a bilayer, as well as the simulation files analysed by the machine learning models. Code source for the ML algorithm can be found on Github. The training files are named gel.gro and fluid.gro. They respectively correspond to the final frame of the systems simulated at \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    358K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       ". All other files are the files analysed by the machine learning models. System composition: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    DPPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules: 212 with 130 atoms each \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules: 29,826 with 3 atoms each Simulation box dimensions (approx.): 8 x 8 x 20 nm Simulation details: \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Software: Gromacs\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " (\n",
       "<mark class=\"entity\" style=\"background: #orange; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    v. 2016.4\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTVERS</span>\n",
       "</mark>\n",
       ") \n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Forcefield: Charmm36\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       " (\n",
       "<mark class=\"entity\" style=\"background: #orange; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    v. June 2015\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTVERS</span>\n",
       "</mark>\n",
       ") - \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ": \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    TIP3P\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " Thermostat: Nose-hoover (0.4ps, 2 groups) Barostat: Parrinello-Rahman semi-isotropic (2.0ps, 1.0 bar on each axis, 4.5e-5 bar-1) \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Duration: 50 ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION - claude_3_7_sonnet:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">212 \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    DPPC&lt;/MOL&gt; Molecules bilayer in pure \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "Water&lt;/MOL&gt;, simulated at temperatures ranging from \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K&lt;/TEMP&gt; to \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       "358K&lt;/TEMP&gt;\\nPublication: A machine learning study of the two states model for lipid bilayer phase transitions Published on: 12 August 2020 Journal: Phys. Chem. Chem. Phys., 2020, DOI: 10.1039/D0CP02058C Description: Simulation files used to train our machine learning algorithm to identify the thermodynamic phase of individual lipid molecules in a bilayer, as well as the simulation files analysed by the machine learning models. Code source for the ML algorithm can be found on Github. The training files are named gel.gro and fluid.gro. They respectively correspond to the final frame of the systems simulated at \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K&lt;/TEMP&gt; and \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       "358K&lt;/TEMP&gt;. All other files are the files analysed by the machine learning models. System composition: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    DPPC&lt;/MOL&gt; molecules: 212 with 130 atoms each \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "Water&lt;/MOL&gt; molecules: 29,826 with 3 atoms each Simulation box dimensions (approx.): 8 x 8 x 20 nm Simulation details: Software: &lt;SOFTNAME&gt;Gromacs&lt;/SOFTNAME&gt; (\n",
       "<mark class=\"entity\" style=\"background: #orange; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    v. 2016.4&lt;/SOFTVERS&gt;) Forcefield: &lt;FFM&gt;Charmm36&lt;/FFM&gt; (\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTVERS</span>\n",
       "</mark>\n",
       "v. June 2015&lt;/SOFTVERS&gt;) - \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water&lt;/MOL&gt;: \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "TIP3P&lt;/MOL&gt; Thermostat: Nose-hoover (0.4ps, 2 groups) Barostat: Parrinello-Rahman semi-isotropic (2.0ps, 1.0 bar on each axis, 4.5e-5 bar-1) Duration: &lt;STIME&gt;50 ns&lt;/STIME&gt;</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION - mistral_saba_24b:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    212\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       " \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    DPPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " Molecules bilayer in pure \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       ", simulated at temperatures ranging from \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       " to \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    358K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       "\\nPublication: A machine learning study of the two states model for lipid bilayer phase transitions Published on: 12 August 2020 Journal: Phys. Chem. Chem. Phys., 2020, DOI: 10.1039/D0CP02058C Description: Simulation files used to train our machine learning algorithm to identify the thermodynamic phase of individual lipid molecules in a bilayer, as well as the simulation files analysed by the machine learning models. Code source for the ML algorithm can be found on Github. The training files are named gel.gro and fluid.gro. They respectively correspond to the final frame of the systems simulated at \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    288K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       " and \n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    358K\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       ". All other files are the files analysed by the machine learning models. System composition: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    DPPC\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules: 212 with 130 atoms each \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Water\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " molecules: 29,826 with 3 atoms each Simulation box dimensions (approx.): 8 x 8 x 20 nm Simulation details: \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Software: Gromacs\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " (\n",
       "<mark class=\"entity\" style=\"background: #orange; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    v. 2016.4\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTVERS</span>\n",
       "</mark>\n",
       ") \n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    Forcefield: Charmm36\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       " (\n",
       "<mark class=\"entity\" style=\"background: #cdb4db; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    v. June 2015\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">FFM</span>\n",
       "</mark>\n",
       ") - Water: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    TIP3P\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " &lt;FFM&gt;Thermostat: Nose-hoover (0.4ps, 2 groups) Barostat: Parrinello-Rahman semi-isotropic (2.0ps, 1.0 bar on each axis, 4.5e-5 bar-1) Duration: \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    50 ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ai_data = [\"../ai_annotations/chatgpt4o_zenodo_3950029.txt\", \"../ai_annotations/gemini_2_0_flash_zenodo_3950029.txt\", \"../ai_annotations/claude_3_7_sonnet_zenodo_3950029.txt\", \"../ai_annotations/mistral_saba_24b_zenodo_3950029.txt\"]\n",
    "model_names = [\"chatgpt4o\", \"gemini_2_0_flash\", \"claude_3_7_sonnet\", \"mistral_saba_24b\"]\n",
    "\n",
    "for i in range(len(ai_data)):\n",
    "    converted_ai_data = convert_annotated_text_to_spacy_format(ai_data[i])\n",
    "    colors = {\n",
    "        \"TEMP\": \"#ffb3ba\",\n",
    "        \"SOFTNAME\": \"#ffffba\",\n",
    "        \"SOFTVERS\": \"#orange\",\n",
    "        \"STIME\": \"#baffc9\",\n",
    "        \"MOL\": \"#bae1ff\",\n",
    "        \"FFM\": \"#cdb4db\",\n",
    "    }\n",
    "    options = {\"colors\": colors}\n",
    "\n",
    "    print(f\"\\nAI ANNOTATION - {model_names[i]}:\")\n",
    "    displacy.render(converted_ai_data, style=\"ent\", manual=True, options=options)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Time to test out one-shot prompting with different models**\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```txt\n",
    "You are an expert at NER, you are meticulous and precise with extensive molecular dynamics and biology knowledge. You are given a scientific abstract or dataset description related to molecular dynamics simulations. Your task is to identify and annotate specific named entities relevant to simulation setup and analysis.\n",
    "\n",
    "Entity Labels to Extract:\n",
    "<SOFTNAME>: Software used in the simulation or analysis (e.g., Gromacs, AMBER, VMD, CHARMM-GUI)\n",
    "<SOFTVERS>: Version number associated with the software (e.g., v. 2016.4, 5.0.3, June 2015)\n",
    "<MOL>: Molecules, proteins, lipids, water models, or molecular complexes involved (e.g., DPPC, water, TIP3P, GTP, KRas4B, tetramer IV)\n",
    "<STIME>: Duration of the simulation (e.g., 50 ns, 200ns, 5 µs)\n",
    "<TEMP>: Temperature used in the simulation (e.g., 300 K, 288K, 358K)\n",
    "<FFM>: Force fields used in the simulation (e.g., Charmm36, AMBER, MARTINI)\n",
    "\n",
    "Expected Output Format:\n",
    "Wrap each identified entity with its label using this format:\n",
    "<LABEL>entity<LABEL>\n",
    "\n",
    "Example:\n",
    "Simulations were performed using <SOFTNAME>Gromacs<SOFTNAME> with the <FFM>Charmm36<FFM> force field at <TEMP>300 K<TEMP> for <STIME>50 ns<STIME>.\n",
    "\n",
    "Input:\n",
    "\"Assessment of mutation probabilities of KRAS G12 missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling: Datasets.\\nDatasets related to the publication [1]. Including: KRAS G12X mutations derived from COSMIC v.79 [http://cancer.sanger.ac.uk/cosmic/] (KRAS G12X mut COSMICv79..xlsx) RMSFs (300-2000ns) of GDP-systems (300 2000rmsf GDP systems RAW AVG SE.xlsx) RMSFs (300-2000ns) of GTP-systems (300 2000RMSF GTP systems RAW AVG SE.xlsx) PyInteraph analysis data for salt-bridges and hydrophobic clusters (.dat files for each system in the PyInteraph data.zip-file) Backbone trajectories for each system (residues 4-164; frames for every 1ns). Instead of using GROMACS, we used NAMD also.”\n",
    "\n",
    "Output:\n",
    "Return the input text with the appropriate entities labeled inline as shown above.\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **REAL ANNOTATION**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Assessment of mutation probabilities of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " G12 missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling: Datasets.<br>Datasets related to the publication [1]. Including: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " G12X mutations derived from COSMIC v.79 [http://cancer.sanger.ac.uk/cosmic/] (KRAS G12X mut COSMICv79..xlsx) RMSFs (300-2000ns) of GDP-systems (300 2000rmsf GDP systems RAW AVG SE.xlsx) RMSFs (300-2000ns) of GTP-systems (300 2000RMSF GTP systems RAW AVG SE.xlsx) \n",
       "<mark class=\"entity\" style=\"background: #ddd; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    PyInteraph\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFT</span>\n",
       "</mark>\n",
       " analysis data for salt-bridges and hydrophobic clusters (.dat files for each system in the PyInteraph data.zip-file) Backbone trajectories for each system (residues 4-164; frames for every 1ns). Last number (e.g. 1) refers to the replica of the simulated system. backbone 4-164.gro/.pdb/.tpr -files (resid 4-164) [1] Pantsar T et al. Assessment of mutation probabilities of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " G12 missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling. PLoS Comput Biol Submitted (2018)</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "real_data = \"../annotations/zenodo_1346073.json\"\n",
    "\n",
    "# Convert the annotations.\n",
    "converted_data = convert_annotations(real_data)\n",
    "\n",
    "colors = {\n",
    "    \"TEMP\": \"#ffb3ba\",\n",
    "    \"SOFTNAME\": \"#ffffba\",\n",
    "    \"SOFTVERS\": \"#orange\",\n",
    "    \"STIME\": \"#baffc9\",\n",
    "    \"MOL\": \"#bae1ff\",\n",
    "    \"FFM\": \"#cdb4db\",\n",
    "}\n",
    "options = {\"colors\": colors}\n",
    "\n",
    "# You now have a list of annotation dictionaries. For visualization with displaCy,\n",
    "# you can serve a single annotated text or multiple.\n",
    "displacy.render(converted_data, style=\"ent\", manual=True, options=options)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **A.I. ANNOTATIONS**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Assessment of mutation probabilities of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS&lt;MOl&gt; G12 missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling: Datasets.\\nDatasets related to the publication [1]. Including: \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "KRAS&lt;MOl&gt; G12X mutations derived from COSMIC v. &lt;SOFTVERS&gt;79 [http://cancer.sanger.ac.uk/cosmic/] (\n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS&lt;MOl&gt; G12X mut COSMICv79..xlsx) RMSFs (&lt;SOFTNAME&gt;NAMD&lt;SOFTNAME&gt; for &lt;STIME&gt;300&lt;STIME&gt;-&lt;STIME&gt;2000ns&lt;STIME&gt;) of \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "GDP system\n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    -systems (&lt;SOFTNAME&gt;NAMD&lt;SOFTNAME&gt; for &lt;STIME&gt;300&lt;STIME&gt;-&lt;STIME&gt;2000ns&lt;STIME&gt; rmsf \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "GDP system&lt;s|MOL&gt; RAW AVG SE.xlsx) RMSFs (\n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    NAMD\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " for \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "-\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    2000ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       ") of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP system&lt;s|MOL&gt;-systems (&lt;SOFTNAME&gt;NAMD&lt;SOFTNAME&gt; for &lt;STIME&gt;300&lt;STIME&gt;-&lt;STIME&gt;2000ns&lt;STIME&gt; RMSF \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "GTP system&lt;s|MOL&gt; RAW AVG SE.xlsx) PyInteraph analysis data for salt-bridges and hydrophobic clusters (.dat files for each system in the PyInteraph data.zip-file) Backbone trajectories for each system (residues 4-164; frames for every \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    1ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "). Instead of using \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       ", we used \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    NAMD\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " also.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ai_data = \"../ai_annotations/llama_3_1_b_annotation.txt\"\n",
    "\n",
    "converted_ai_data = convert_annotated_text_to_spacy_format(ai_data)\n",
    "\n",
    "colors = {\n",
    "    \"TEMP\": \"#ffb3ba\",\n",
    "    \"SOFTNAME\": \"#ffffba\",\n",
    "    \"SOFTVERS\": \"#orange\",\n",
    "    \"STIME\": \"#baffc9\",\n",
    "    \"MOL\": \"#bae1ff\",\n",
    "    \"FFM\": \"#cdb4db\",\n",
    "}\n",
    "options = {\"colors\": colors}\n",
    "\n",
    "print(\"\\nAI ANNOTATION:\")\n",
    "displacy.render(converted_ai_data, style=\"ent\", manual=True, options=options)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Assessment of mutation probabilities of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS G12&lt;\tMOL&gt; missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling: Datasets.\\nDatasets related to the publication [1]. Including: \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "KRAS G12X&lt;\tMOL&gt; mutations derived from COSMIC v.79 [http://cancer.sanger.ac.uk/cosmic/] (KRAS G12X mut COSMICv79..xlsx) RMSFs (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300-2000ns&lt;\tSTIME&gt;) of &lt;MOL&gt;GDP-systems&lt;\tMOL&gt; (\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "300&lt;\tSTIME&gt; \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    2000&lt;\tSTIME&gt;rmsf GDP systems RAW AVG SE.xlsx) RMSFs (\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "300-2000ns&lt;\tSTIME&gt;) of &lt;MOL&gt;GTP-systems&lt;\tMOL&gt; (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300&lt;\tSTIME&gt; \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "2000&lt;\tSTIME&gt;RMSF GTP systems RAW AVG SE.xlsx) PyInteraph analysis data for salt-bridges and hydrophobic clusters (.dat files for each system in the PyInteraph data.zip-file) Backbone trajectories for each system (residues 4-164; frames for every &lt;STIME&gt;1ns&lt;\tSTIME&gt;). Instead of using \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS&lt;\tSOFTNAME&gt;, we used \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       "NAMD&lt;\tSOFTNAME&gt; also.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ai_data = \"../ai_annotations/deepseek_qwen.txt\"\n",
    "\n",
    "converted_ai_data = convert_annotated_text_to_spacy_format(ai_data)\n",
    "\n",
    "colors = {\n",
    "    \"TEMP\": \"#ffb3ba\",\n",
    "    \"SOFTNAME\": \"#ffffba\",\n",
    "    \"SOFTVERS\": \"#orange\",\n",
    "    \"STIME\": \"#baffc9\",\n",
    "    \"MOL\": \"#bae1ff\",\n",
    "    \"FFM\": \"#cdb4db\",\n",
    "}\n",
    "options = {\"colors\": colors}\n",
    "\n",
    "print(\"\\nAI ANNOTATION:\")\n",
    "displacy.render(converted_ai_data, style=\"ent\", manual=True, options=options)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">ssessment of mutation probabilities of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS G12&lt;/MOL&gt; missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling: Datasets.\\nDatasets related to the publication [1]. Including: \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "KRAS G12X&lt;/MOL&gt; mutations derived from \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    COSMIC&lt;/SOFTNAME&gt; v.&lt;SOFTVERS&gt;79&lt;/SOFTVERS&gt; [http://cancer.sanger.ac.uk/cosmic/] (\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       "KRAS G12X mut COSMICv79..xlsx&lt;/SOFTNAME&gt;) \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    RMSFs&lt;/STIME&gt; (&lt;TEMP&gt;300-2000ns&lt;/TEMP&gt;) of &lt;MOL&gt;GDP-systems&lt;/MOL&gt; (&lt;TEMP&gt;300 2000rmsf GDP systems RAW AVG SE.xlsx&lt;/TEMP&gt;) \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "RMSFs&lt;/STIME&gt; (\n",
       "<mark class=\"entity\" style=\"background: #ffb3ba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300-2000ns&lt;/TEMP&gt;) of &lt;MOL&gt;GTP-systems&lt;/MOL&gt; (\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">TEMP</span>\n",
       "</mark>\n",
       "300 2000RMSF GTP systems RAW AVG SE.xlsx&lt;/TEMP&gt;) \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    PyInteraph&lt;/SOFTNAME&gt; analysis data for salt-bridges and hydrophobic clusters (.dat files for each system in the \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       "PyInteraph&lt;/SOFTNAME&gt; data.zip-file) Backbone trajectories for each system (residues 4-164; frames for every 1ns). Instead of using \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS&lt;/SOFTNAME&gt;, we used \n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       "NAMD&lt;/SOFTNAME&gt; also.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ai_data = \"../ai_annotations/gemma2.txt\"\n",
    "\n",
    "converted_ai_data = convert_annotated_text_to_spacy_format(ai_data)\n",
    "\n",
    "colors = {\n",
    "    \"TEMP\": \"#ffb3ba\",\n",
    "    \"SOFTNAME\": \"#ffffba\",\n",
    "    \"SOFTVERS\": \"#orange\",\n",
    "    \"STIME\": \"#baffc9\",\n",
    "    \"MOL\": \"#bae1ff\",\n",
    "    \"FFM\": \"#cdb4db\",\n",
    "}\n",
    "options = {\"colors\": colors}\n",
    "\n",
    "print(\"\\nAI ANNOTATION:\")\n",
    "displacy.render(converted_ai_data, style=\"ent\", manual=True, options=options)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I tried telling the model to remove the backslashes from the labels but it didn't change much. It would keep adding them in the label."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Assessment of mutation probabilities of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS G12\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling: Datasets.<br>Datasets related to the publication [1]. Including: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS G12X\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " mutations derived from COSMIC \n",
       "<mark class=\"entity\" style=\"background: #orange; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    v.79\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTVERS</span>\n",
       "</mark>\n",
       " [http://cancer.sanger.ac.uk/cosmic/] (KRAS G12X mut COSMICv79..xlsx) RMSFs (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300-2000ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       ") of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GDP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-systems (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300 2000ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       " RMSF GDP systems RAW AVG SE.xlsx) RMSFs (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300-2000ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       ") of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-systems (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300 2000ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       " RMSF GTP systems RAW AVG SE.xlsx) \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    PyInteraph\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " analysis data for salt-bridges and hydrophobic clusters (.dat files for each system in the PyInteraph data.zip-file) Backbone trajectories for each system (residues 4-164; frames for every \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    1ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "). Instead of using \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       ", we used \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    NAMD\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " also.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ai_data = \"../ai_annotations/chatgpt4o.txt\"\n",
    "\n",
    "converted_ai_data = convert_annotated_text_to_spacy_format(ai_data)\n",
    "\n",
    "colors = {\n",
    "    \"TEMP\": \"#ffb3ba\",\n",
    "    \"SOFTNAME\": \"#ffffba\",\n",
    "    \"SOFTVERS\": \"#orange\",\n",
    "    \"STIME\": \"#baffc9\",\n",
    "    \"MOL\": \"#bae1ff\",\n",
    "    \"FFM\": \"#cdb4db\",\n",
    "}\n",
    "options = {\"colors\": colors}\n",
    "\n",
    "print(\"\\nAI ANNOTATION:\")\n",
    "displacy.render(converted_ai_data, style=\"ent\", manual=True, options=options)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Assessment of mutation probabilities of KRAS G12 missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling: Datasets.\\nDatasets related to the publication [1]. Including: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS G12X\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " mutations derived from COSMIC v.79 [http://cancer.sanger.ac.uk/cosmic/] (\n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS G12X\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " mut COSMICv79..xlsx) RMSFs (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300-2000ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       ") of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GDP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-systems (300 2000rmsf \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GDP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " systems RAW AVG SE.xlsx) RMSFs (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300-2000ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       ") of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-systems (300 2000RMSF \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " systems RAW AVG SE.xlsx) PyInteraph analysis data for salt-bridges and hydrophobic clusters (.dat files for each system in the PyInteraph data.zip-file) Backbone trajectories for each system (residues 4-164; frames for every 1ns). Instead of using \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       ", we used \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    NAMD\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " also.</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ai_data = \"../ai_annotations/gemini_2_0_flash.txt\"\n",
    "\n",
    "converted_ai_data = convert_annotated_text_to_spacy_format(ai_data)\n",
    "\n",
    "colors = {\n",
    "    \"TEMP\": \"#ffb3ba\",\n",
    "    \"SOFTNAME\": \"#ffffba\",\n",
    "    \"SOFTVERS\": \"#orange\",\n",
    "    \"STIME\": \"#baffc9\",\n",
    "    \"MOL\": \"#bae1ff\",\n",
    "    \"FFM\": \"#cdb4db\",\n",
    "}\n",
    "options = {\"colors\": colors}\n",
    "\n",
    "print(\"\\nAI ANNOTATION:\")\n",
    "displacy.render(converted_ai_data, style=\"ent\", manual=True, options=options)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Assessment of mutation probabilities of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS G12\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling: Datasets.\\nDatasets 1  related to the publication [1]. Including: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS G12X\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " mutations derived from COSMIC v.79 [http://cancer.sanger.ac.uk/cosmic/] (\n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS G12X\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " mut COSMICv79..xlsx) RMSFs (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300-2000ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       ") of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GDP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-systems (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300-2000ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       " rmsf \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GDP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " systems RAW AVG SE.xlsx) RMSFs (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300-2000ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       ") of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-systems (\n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    300-2000ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       " RMSF \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " systems RAW AVG SE.xlsx) \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    PyInteraph\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " analysis data for salt-bridges and hydrophobic clusters (.dat files for each system in the \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    PyInteraph\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " data.zip-file) Backbone trajectories for each system (residues 4-164; frames for every \n",
       "<mark class=\"entity\" style=\"background: #baffc9; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    1ns\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">STIME</span>\n",
       "</mark>\n",
       "). Instead of using \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       ", we used \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    NAMD\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " also</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ai_data = \"../ai_annotations/gemini_2_0_flash_thinking.txt\"\n",
    "\n",
    "converted_ai_data = convert_annotated_text_to_spacy_format(ai_data)\n",
    "\n",
    "colors = {\n",
    "    \"TEMP\": \"#ffb3ba\",\n",
    "    \"SOFTNAME\": \"#ffffba\",\n",
    "    \"SOFTVERS\": \"#orange\",\n",
    "    \"STIME\": \"#baffc9\",\n",
    "    \"MOL\": \"#bae1ff\",\n",
    "    \"FFM\": \"#cdb4db\",\n",
    "}\n",
    "options = {\"colors\": colors}\n",
    "\n",
    "print(\"\\nAI ANNOTATION:\")\n",
    "displacy.render(converted_ai_data, style=\"ent\", manual=True, options=options)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here I had to specify to not use backslashes in the labels. It was doing a html format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "AI ANNOTATION:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<span class=\"tex2jax_ignore\"><div class=\"entities\" style=\"line-height: 2.5; direction: ltr\">Assessment of mutation probabilities of KRAS G12 missense mutants and their long-time scale dynamics by atomistic molecular simulations and Markov state modeling: Datasets.<br>Datasets related to the publication [1]. Including: \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS G12X\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " mutations derived from COSMIC v.79 [http://cancer.sanger.ac.uk/cosmic/] (\n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    KRAS G12X\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " mut COSMICv79..xlsx) RMSFs (300-2000ns) of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GDP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-systems (300 2000rmsf \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GDP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " systems RAW AVG SE.xlsx) RMSFs (300-2000ns) of \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       "-systems (300 2000RMSF \n",
       "<mark class=\"entity\" style=\"background: #bae1ff; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GTP\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">MOL</span>\n",
       "</mark>\n",
       " systems RAW AVG SE.xlsx) PyInteraph analysis data for salt-bridges and hydrophobic clusters (.dat files for each system in the PyInteraph data.zip-file) Backbone trajectories for each system (residues 4-164; frames for every 1ns). Instead of using \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    GROMACS\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       ", we used \n",
       "<mark class=\"entity\" style=\"background: #ffffba; padding: 0.45em 0.6em; margin: 0 0.25em; line-height: 1; border-radius: 0.35em;\">\n",
       "    NAMD\n",
       "    <span style=\"font-size: 0.8em; font-weight: bold; line-height: 1; border-radius: 0.35em; vertical-align: middle; margin-left: 0.5rem\">SOFTNAME</span>\n",
       "</mark>\n",
       " also</div></span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ai_data = \"../ai_annotations/claude_3_7_sonnet.txt\"\n",
    "\n",
    "converted_ai_data = convert_annotated_text_to_spacy_format(ai_data)\n",
    "\n",
    "colors = {\n",
    "    \"TEMP\": \"#ffb3ba\",\n",
    "    \"SOFTNAME\": \"#ffffba\",\n",
    "    \"SOFTVERS\": \"#orange\",\n",
    "    \"STIME\": \"#baffc9\",\n",
    "    \"MOL\": \"#bae1ff\",\n",
    "    \"FFM\": \"#cdb4db\",\n",
    "}\n",
    "options = {\"colors\": colors}\n",
    "\n",
    "print(\"\\nAI ANNOTATION:\")\n",
    "displacy.render(converted_ai_data, style=\"ent\", manual=True, options=options)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
